# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/gtest/TestDeviceInputTrack.cpp
# Commit: 8cb7b8231c98
# Full Hash: 8cb7b8231c980aa7830fad2ecc34cb128417d511
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2022-01-28 04:41:16
# Regressor Bug: 1742655
# File Overlap Count: 1
# Description:
#   Bug 1742655 - Open/Close audio input via NativeInputTrack r=pehrsons
#   
#   Move the audio-opening check logic and the creation of NativeInputTrack
#   in one place instead of seperating them in different functions. This
#   reduces the complexity of audio-opening logic in MediaTrackGraph and
# ==============================================================================

diff -r a1def130c105 -r 8cb7b8231c98 dom/media/gtest/TestDeviceInputTrack.cpp
--- a/dom/media/gtest/TestDeviceInputTrack.cpp	Thu Jan 27 22:29:34 2022 +0000
+++ b/dom/media/gtest/TestDeviceInputTrack.cpp	Thu Jan 27 22:29:35 2022 +0000
@@ -84,8 +84,10 @@
   const PrincipalHandle testPrincipal =
       MakePrincipalHandle(nsContentUtils::GetSystemPrincipal());
 
-  RefPtr<NativeInputTrack> track =
-      NativeInputTrack::Create(mGraph.get(), deviceId, testPrincipal);
+  auto r = NativeInputTrack::OpenAudio(mGraph.get(), deviceId, testPrincipal,
+                                       nullptr);
+  ASSERT_TRUE(r.isOk());
+  RefPtr<NativeInputTrack> track = r.unwrap();
 
   generator.GenerateInterleaved(buffer.Elements(), nrFrames);
   track->NotifyInputData(mGraph.get(), buffer.Elements(), nrFrames, mRate,
@@ -121,6 +123,24 @@
     EXPECT_EQ(chunk.mPrincipalHandle, testPrincipal);
   }
 
-  track->Destroy();
-  mGraph->RemoveTrackGraphThread(track);
+  NativeInputTrack::CloseAudio(std::move(track), nullptr);
 }
+
+TEST_F(TestDeviceInputTrack, OpenTwiceWithoutCloseFirst) {
+  const CubebUtils::AudioDeviceID deviceId1 = (void*)1;
+  const CubebUtils::AudioDeviceID deviceId2 = (void*)2;
+
+  const PrincipalHandle testPrincipal =
+      MakePrincipalHandle(nsContentUtils::GetSystemPrincipal());
+
+  auto r1 = NativeInputTrack::OpenAudio(mGraph.get(), deviceId1, testPrincipal,
+                                        nullptr);
+  ASSERT_TRUE(r1.isOk());
+  RefPtr<NativeInputTrack> track1 = r1.unwrap();
+
+  auto r2 = NativeInputTrack::OpenAudio(mGraph.get(), deviceId2, testPrincipal,
+                                        nullptr);
+  EXPECT_TRUE(r2.isErr());
+
+  NativeInputTrack::CloseAudio(std::move(track1), nullptr);
+}